### PR TITLE
stdlib/SparseArrays: fix scalar setindex! for vector eltype

### LINKED ISSUE #29034: SparseMatrixCSC as a general sparse container: Setindex!(…) not working for SparseMatrixCSC with vector elements (v0.7)
In my application I have a sparse matrix, but every element of this matrix is a discretised function. If I try to use `SparseMatrixCSC{Array{Float64,1},Int64}` as the container, `setindex!(...)` in 0.7 throws a stackoverflow:

```julia
julia> sa=sparse([1],[1],[Vector{Float64}(undef,3)],3,3)
3×3 SparseMatrixCSC{Array{Float64,1},Int64} with 1 stored entry:
  [1, 1]  =  [6.93279e-310, 6.93279e-310, 6.93279e-310]

julia> sa[1,1] = [1.0,2.0,3.0]
StackOverflowError:
Stacktrace:
 [1] setindex!(::SparseMatrixCSC{Array{Int64,1},Int64}, ::Array{Int64,1}, ::Int64, ::Int64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2540 (repeats 80000 times)
```
This can be solved by using broadcasted `setindex!(...)`, but it does not solve the problem adding new elements to the array:
```julia
julia> sa[1,1] .= [1.0,2.0,3.0]
3-element Array{Float64,1}:
 1.0
 2.0
 3.0

julia> sa[2,2] = [1.0,2.0,3.0]
StackOverflowError:
Stacktrace:
 [1] setindex!(::SparseMatrixCSC{Array{Float64,1},Int64}, ::Array{Float64,1}, ::Int64, ::Int64) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v0.7/SparseArrays/src/sparsematrix.jl:2540 (repeats 80000 times)
```

I can work around it using custom structures, but this has worked in 0.6.x.

The version producing the bug
```julia
julia> versioninfo()
Julia Version 0.7.0
Commit a4cb80f3ed (2018-08-08 06:46 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```

### PR DESCRIPTION
**Pull request by stev47**
fixes #29034 

There might be more elegant solutions, I'm open to suggestions.

### CODE REVIEW

**Comment by stev47** [comment]
Appveyor failure seems unrelated

**Comment by ViralBShah** [comment]
Can we merge this?

File: stdlib/SparseArrays/src/sparsematrix.jl [hunk 1]
```diff
@@ -2565,7 +2569,8 @@ end
 _to_same_csc(::SparseMatrixCSC{Tv, Ti}, V::AbstractMatrix, I...) where {Tv,Ti} = convert(SparseMatrixCSC{Tv,Ti}, V)
 _to_same_csc(::SparseMatrixCSC{Tv, Ti}, V::AbstractVector, I...) where {Tv,Ti} = convert(SparseMatrixCSC{Tv,Ti}, reshape(V, map(length, I)))
 
-setindex!(A::SparseMatrixCSC{Tv}, B::AbstractVecOrMat, I::Integer, J::Integer) where {Tv} = setindex!(A, convert(Tv, B), I, J)
+setindex!(A::SparseMatrixCSC{Tv}, B::AbstractVecOrMat, I::Integer, J::Integer) where {Tv} = _setindex_scalar!(A, B, I, J)
```
Comment (mbauman) [inline stdlib/SparseArrays/src/sparsematrix.jl]: I was hoping we might be able to remove this method entirely, but it looks like it's disambiguating against the method that immediately follows.  Using a helper function like this is a good solution.

**Review by mbauman — APPROVED** [review]

**Comment by mortenpi** [comment]
Should this be backported to 1.0 too? The issue is also present in 1.0.3.

**Comment by mbauman** [comment]
Yes, I'd support backporting it — it's a fairly major bugfix.

**Comment by mschauer** [comment]
Ping. I run into this, I believe on Julia Version 1.1.0:

```
Compute guiding term:
ERROR: LoadError: StackOverflowError:
Stacktrace:
 [1] setindex!(::SparseMatrixCSC{SArray{Tuple{2,2},Float64,2,4},Int64}, ::SArray{Tuple{2,2},Float64,2,4}, ::Int64, ::Int64) at /Users/osx/buildbot/slave/package_osx64/build/usr/share/julia/stdlib/v1.1/SparseArrays/src/sparsematrix.jl:2568 (repeats 80000 times)
```

**Comment by stev47** [comment]
probably because this pull request has not yet been backported

**Pull request merged**